citylearn/electric_vehicle_charger.py

```diff
@@ -0,0 +1,374 @@
+from citylearn.base import Environment
+
+from citylearn.electric_vehicle import electric_vehicle
+
+ZERO_DIVISION_CAPACITY = 0.00001
+
+
+class Charger(Environment):
+    def __init__(
+            self,
+            nominal_power: float,
+            efficiency: float = None,
+            charger_id: str = None,
+            charger_type: int = None,
+            max_charging_power: float = 50.0,
+            min_charging_power: float = 0.0,
+            max_discharging_power: float = 50.0,
+            min_discharging_power: float = 0.0,
+            charge_efficiency_curve: Dict[float, float] = None,
+            discharge_efficiency_curve: Dict[float, float] = None,
+            image_path: str = None,
+            connected_ev: electric_vehicle = None, incoming_ev: electric_vehicle = None,
+            **kwargs
+    ):
+        r"""Initializes the `Electric Vehicle Charger` class with the given attributes.
+
+        Parameters
+        ----------
+        charger_id: str
+            Id through which the charger is uniquely identified in the system
+        charger_type: int
+            Either private (0) or public (1) charger
+        max_charging_power : float, default 50
+            Maximum charging power in kW.
+        min_charging_power : float, default 0
+            Minimum charging power in kW.
+        max_discharging_power : float, default 50
+            Maximum discharging power in kW.
+        min_discharging_power : float, default 0
+            Minimum discharging power in kW.
+        charge_efficiency_curve : dict, default {3.6: 0.95, 7.2: 0.97, 22: 0.98, 50: 0.98}
+            Efficiency curve for charging containing power levels and corresponding efficiency values.
+        discharge_efficiency_curve : dict, default {3.6: 0.95, 7.2: 0.97, 22: 0.98, 50: 0.98}
+            Efficiency curve for discharging containing power levels and corresponding efficiency values.
+        max_connected_cars : int, default 1
+            Maximum number of cars that can be connected to the charger simultaneously.
+
+        Other Parameters
+        ----------------
+        **kwargs : dict
+            Other keyword arguments used to initialize super classes.
+        """
+
+        self.nominal_power = nominal_power
+        self.efficiency = efficiency
+        self.charger_id = charger_id
+        self.charger_type = charger_type
+        self.max_charging_power = max_charging_power
+        self.min_charging_power = min_charging_power
+        self.max_discharging_power = max_discharging_power
+        self.min_discharging_power = min_discharging_power
+        self.charge_efficiency_curve = charge_efficiency_curve or {3.6: 0.95, 7.2: 0.97, 22: 0.98, 50: 0.98}
+        self.discharge_efficiency_curve = discharge_efficiency_curve or {3.6: 0.95, 7.2: 0.97, 22: 0.98, 50: 0.98}
+        self.image_path = image_path
+        self.connected_ev = connected_ev or None
+        self.incoming_ev = incoming_ev or None
```

Comment:
This is not a good way to define defaults because if the value is changed to None after calling the __init__ function, there will be no default to fall back on. Use a setter method instead.

Reply:
Please Double-check the changes made.